File: main/course-schedule-ii/course-schedule-ii.py
import enum


class Solution:
    @staticmethod
    def findOrder(numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        """Finds an order that courses may be taken. Returns [] on failure."""
        adj = build_graph(numCourses, prerequisites)
        out = reverse_topological_sort(adj)
        return [] if out is None else out


def build_graph(order, edges):
    """
    Builds an adjacency list for a directed graph with the given number of
    vertices (order) and directed edges u -> v represented (u, v).
    """
    adj = [[] for _ in range(order)]

    for src, dest in edges:
        adj[src].append(dest)

    return adj


@enum.unique
class Color(enum.Enum):
    """Visitation states for directed graph traversal."""
    WHITE = enum.auto()  # unvisited
    GRAY = enum.auto()   # visited but not yet fully explored
    BLACK = enum.auto()  # visited and fully explored


def reverse_topological_sort(adj):
    """
    Computes a reverse topological sort of a directed graph via recursive DFS,
    given the graph's adjacency list. Returns None if the graph is cyclic.
    """
    vis = [Color.WHITE] * len(adj)
    out = []

    def has_cycle_from(src):
        if vis[src] is Color.BLACK:
            return False
        if vis[src] is Color.GRAY:
            return True
        assert vis[src] is Color.WHITE, 'invalid vistation state'

        vis[src] = Color.GRAY
        if any(has_cycle_from(dest) for dest in adj[src]):
            return True
        vis[src] = Color.BLACK

        out.append(src)
        return False

    if any(has_cycle_from(start) for start in range(len(adj))):
        return None

    return out

File: main/course-schedule-ii/course-schedule-ii.java
class Solution {
    public int[] findOrder(final int numCourses, final int[][] prerequisites) {
        final List<List<Integer>> adj = buildAdjacencyList(numCourses,
                                                           prerequisites);

        return reverseTopologicalSort(adj).stream().mapToInt(x -> x).toArray();
    }

    private static List<List<Integer>>
    buildAdjacencyList(int vertexCount, final int[][] edges) {
        // create an empty adjacency list with the correct number of rows
        final List<List<Integer>> adj = new ArrayList<>(vertexCount);
        while (vertexCount-- != 0) adj.add(new ArrayList<>());

        // populate the adjacency list with information about all the edges
        for (final int[] edge : edges) adj.get(edge[0]).add(edge[1]);
        return adj;
    }

    private static List<Integer>
    reverseTopologicalSort(final List<List<Integer>> adj) {
        final Color[] vis = new Color[adj.size()];
        Arrays.fill(vis, Color.WHITE);

        final List<Integer> out = new ArrayList<>(adj.size());

        for (int src = 0; src != adj.size(); ++src) {
            if (hasCycleFrom(adj, vis, out, src)) {
                out.clear();
                break;
            }
        }

        return out;
    }

    private static boolean
    hasCycleFrom(final List<List<Integer>> adj, final Color[] vis,
                 final List<Integer> out, final int src) {
        switch (vis[src]) {
        case WHITE:
            vis[src] = Color.GRAY;

            for (final int dest : adj.get(src))
                if (hasCycleFrom(adj, vis, out, dest)) return true;

            vis[src] = Color.BLACK;
            out.add(src);
            return false;

        case GRAY:
            return true;

        case BLACK:
            return false;
        }

        throw new AssertionError("invalid visitation state");
    }
}

/** Visitation states for traversing a directed graph. */
enum Color {
    /** Unvisited */
    WHITE,

    /** Visited but not yet fully explored */
    GRAY,

    /** Visited and fully explored */
    BLACK
}
